3083 support migration with dois

michbarsinai opened the pull request:
## New Contributors

Welcome! New contributors should at least glance at [CONTRIBUTING.md](/CONTRIBUTING.md), especially the section on pull requests where we encourage you to reach out to other developers before you start coding. Also, please note that we measure code coverage and prefer you write unit tests. Pull requests can still be reviewed without tests or completion of the checklist outlined below. Thanks!

## Related Issues

- connects to #3083: Create Dataset: add support for migrating (with existing DOIs) via native API

## Pull Request Checklist

- [x] Unit [tests][] completed
- [x] Integration [tests][]: None
- [x] Deployment requirements, [SQL updates][], [Solr updates][], etc.: None
- [x] [Documentation][docs] completed
- [x] Merged latest from "develop" [branch][] and resolved conflicts

[tests]: http://guides.dataverse.org/en/latest/developers/testing.html
[SQL updates]: https://github.com/IQSS/dataverse/tree/develop/scripts/database/upgrades
[Solr updates]: https://github.com/IQSS/dataverse/blob/develop/conf/solr/7.2.1/schema.xml
[docs]: http://guides.dataverse.org/en/latest/developers/documentation.html
[branch]: http://guides.dataverse.org/en/latest/developers/branching-strategy.html

pameyer commented:
Ran integration tests on this branch; haven't investigated the failures/errors yet.

Failed tests:   testPrivateUrl(edu.harvard.iq.dataverse.api.DatasetsIT): expected:<200> but was:<400>
  testMalformedFacetQueryString(edu.harvard.iq.dataverse.api.DataversesIT)
  testJsonParserWithDirectoryLabels(edu.harvard.iq.dataverse.api.FileMetadataIT): JSON path message doesn't match.(..)
  testJsonParserWithDirectoryLabels(edu.harvard.iq.dataverse.api.FileMetadataIT): Expected status code <200> doesn't match actual status code <404>.(..)
  test_008_ReplaceFileAlreadyDeleted(edu.harvard.iq.dataverse.api.FilesIT): Expected status code <204> doesn't match actual status code <400>.(..)
  testReplaceFileBadJson(edu.harvard.iq.dataverse.api.FilesIT): Expected status code <200> doesn't match actual status code <400>.(..)
  testCuratorSendsCommentsToAuthor(edu.harvard.iq.dataverse.api.InReviewWorkflowIT): JSON path data.inReview doesn't match.(..)
  testCreateDatasetPublishDestroy(edu.harvard.iq.dataverse.api.SwordIT): Expected status code <200> doesn't match actual status code <400>.(..)
  testDeleteFiles(edu.harvard.iq.dataverse.api.SwordIT): Expected status code <204> doesn't match actual status code <400>.(..)

Tests in error: 
  test_006_ReplaceFileGood(edu.harvard.iq.dataverse.api.FilesIT): Cannot get property 'files' on null object
  testForceReplace(edu.harvard.iq.dataverse.api.FilesIT): Cannot get property 'files' on null object

Tests run: 69, Failures: 9, Errors: 2, Skipped: 0

pameyer commented:
First pass run-through looks good!

Will continue testing, but a few questions so far:
- `storageIdentifier` in JSON key looks like it accepts user-specified values.  Is there a scheme I should be using to generate them, or just make sure they're unique?
- `prt:auth:id` struck me as slightly non-intuitive, since I've usually seen it as `prt:auth/id`.  What's the thinking here?

michbarsinai commented:
* re: Storage identifier - Unsure. I'd ask Gustavo or Leonid
* re: Thinking behind PID - mostly over-thinking, really :-). It's related to issue #898 and the way represent PIDs in the system, which could be improved. After this iteration of code review is done, I'll change this to prt:auth/id (already agreed with Gustavo).

> On 26 Apr 2018, at 0:42, pameyer <notifications@github.com> wrote:
> 
> First pass run-through looks good!
> 
> Will continue testing, but a few questions so far:
> 
> storageIdentifier in JSON key looks like it accepts user-specified values. Is there a scheme I should be using to generate them, or just make sure they're unique?
> prt:auth:id struck me as slightly non-intuitive, since I've usually seen it as prt:auth/id. What's the thinking here?
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/IQSS/dataverse/pull/4606#issuecomment-384443682>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AB2UJFtl86yymDoZ9IeB8A9PfrFdrevCks5tsO3GgaJpZM4Te5j5>.
>

pameyer commented:
`filesize` parameter is ignored even when data file directories and manifests are present.

pameyer commented:
travis-ci build failure is a timeout; no issues with `mvn compile` or `mvn package` as of 28c65044370e2d6c60542926fb2728dcc25fab84 when building locally.

pameyer commented:
travis-ci build failure is a timeout; no issues with `mvn compile` or `mvn package` as of 28c65044370e2d6c60542926fb2728dcc25fab84 when building locally.

(comment was from Fri; not sure why it didn't go through then)

pdurbin commented:
@michbarsinai as you may have noticed there are a significant number of files with merge conflicts in this branch (24 files as of 2760c46). Also, previously @pameyer reported failures in the API test suite and I wanted to make sure it's on your radar to make sure those tests are passing. Thanks!

pameyer commented:
travis error looks legit this time - 2760c468e7434c3f062e5e6ea2b0ed184974fc7a does not built with `mvn package` due to an error in  `testparseFiles(edu.harvard.iq.dataverse.util.json.JsonParserTest)`

michbarsinai commented:
I know. There's are some contradicting changes dealing with GlobalIds (and there are more changes in this area coming this week).

I'm actively working on this, once the merge is done I'll move on to the test failures.

> On 21 May 2018, at 17:01, Philip Durbin <notifications@github.com> wrote:
> 
> @michbarsinai <https://github.com/michbarsinai> as you may have noticed there are a significant number of files with merge conflicts in this branch (24 files as of 2760c46 <https://github.com/IQSS/dataverse/commit/2760c468e7434c3f062e5e6ea2b0ed184974fc7a>). Also, previously @pameyer <https://github.com/pameyer> reported failures in the API test suite and I wanted to make sure it's on your radar to make sure those tests are passing. Thanks!
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/IQSS/dataverse/pull/4606#issuecomment-390662250>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AB2UJMYWaxUXcJJTboVhcSfeKBPFqfGSks5t0sjCgaJpZM4Te5j5>.
>

michbarsinai commented:
Merged with develop, all unit tests pass, and the below test-case table works.

Use case | Create | Update | Publish
-- | -- | -- | --
Create DS UI  |OK  |OK  |OK
Create DS API   |OK  |OK  |OK
Import DS good DOI  |OK  |OK  |OK
Import DS bad DOI   | OK  |n/a|   n/a
Import DS good HDL  | OK  |OK  |OK
Import DS bad HDL   | OK  |n/a|   n/a
Import DS Bad PID   | OK  |n/a|   n/a
* OK DOI `doi:10.7910/DVN/L3HN0K`
* bad DOI `doi:10.7910/DVN/XXXXXX`
* OK HDL `hdl:11529/2177350`
* bad HDL `hdl:12529/2177350`
* bad pid `abad:pid`

oscardssmith commented on src/main/java/edu/harvard/iq/dataverse/AbstractGlobalIdServiceBean.java:
```diff
@@ -110,4 +113,18 @@ public DvObject generateIdentifier(DvObject dvObject) {
+     * @param map
+     * @return A HashMap with the same values as {@code map}
+     */
+    protected <T> HashMap<T,T> asHashMap(Map<T,T> map) {
```
This feels fishy to me. Should the internal API's really require HashMap? If so, why??

michbarsinai replied:
It feels fishy, because it is fishy. I've fixed it before, but it slipped back in without me noticing during the last merge. Good catch 👍

michbarsinai replied:
Oh, sorry, I take it back. this *is* the fix, which is why it is protected. The `EZIDService` class (3rd party) takes a `HashMap` as an input. So this is a generic "convert-or-case" adapter. Still a good catch, but solving this seems out of scope. Moved it to `DOIEZIdServiceBean` so it's not part of the public API, though.

oscardssmith commented on src/main/java/edu/harvard/iq/dataverse/DOIEZIdServiceBean.java:
```diff
@@ -124,15 +124,16 @@ public boolean alreadyExists(DvObject dvObject) throws Exception {
         try {
-            ezidService.setMetadata(identifier, metadata);
+            ezidService.setMetadata(identifier,
+                (metadata instanceof HashMap) ? (HashMap)metadata : new HashMap<>(metadata));
```
Given line 132, don't we know that metadata is a HashMap? Given that, why bother with this check?

michbarsinai replied:
Fixed.

oscardssmith commented on src/main/java/edu/harvard/iq/dataverse/DatasetPage.java:
```diff
@@ -3695,15 +3690,15 @@ public void handleLabelsFileUpload(FileUploadEvent event) {
                 uploadStream = file.getInputstream();
             } catch (IOException ioex) {
-                logger.warning("the file " + file.getFileName() + " failed to upload!");
+                logger.log(Level.WARNING, "the file {0} failed to upload!", file.getFileName());
```
Is there a way to use the interpolation without logger.log(Level.WARNING, which is just a bit clunky?

michbarsinai replied:
No. This was discussed before - there a nontrivial runtime cost for composing a string, and if you send a message to the logger and then it's ignored it's just a waste. The form used here is the current "best practice", and was actually created by a NetBeans' code hint.

oscardssmith commented on src/main/java/edu/harvard/iq/dataverse/DatasetServiceBean.java:
```diff
@@ -246,24 +245,34 @@ private String generateIdentifierAsSequentialNumber(Dataset dataset, IdServiceBe
+    }
+    
+    public boolean isIdentifierLocallyUnique(String identifier, Dataset dataset) {
+        boolean u = em.createNamedQuery("Dataset.findByIdentifierAuthorityProtocol")
```
just return, don't assign

michbarsinai replied:
fixed

oscardssmith commented on src/main/java/edu/harvard/iq/dataverse/DataverseRoleServiceBean.java:
```diff
@@ -84,194 +91,195 @@ private IndexResponse indexDefinitionPoint(DvObject definitionPoint) {
+
+        return roles;
+    }
+
```
Are these only whitespace changes?

michbarsinai replied:
Yes.

pameyer commented:
23a1f462ee60d4d376751211c5d086247389d486

Failed tests:   testPrivateUrl(edu.harvard.iq.dataverse.api.DatasetsIT): expected:<200> but was:<400>
  test_008_ReplaceFileAlreadyDeleted(edu.harvard.iq.dataverse.api.FilesIT): Expected status code <204> doesn't match actual status code <400>.(..)
  testReplaceFileBadJson(edu.harvard.iq.dataverse.api.FilesIT): Expected status code <200> doesn't match actual status code <400>.(..)
  testCuratorSendsCommentsToAuthor(edu.harvard.iq.dataverse.api.InReviewWorkflowIT): JSON path data.inReview doesn't match.(..)
  testDatasetThumbnail(edu.harvard.iq.dataverse.api.SearchIT): Expected status code <200> doesn't match actual status code <400>.(..)
  testIdentifier(edu.harvard.iq.dataverse.api.SearchIT): JSON path data.total_count doesn't match.(..)
  testCreateDatasetPublishDestroy(edu.harvard.iq.dataverse.api.SwordIT): Expected status code <200> doesn't match actual status code <400>.(..)
  testDeleteFiles(edu.harvard.iq.dataverse.api.SwordIT): Expected status code <204> doesn't match actual status code <400>.(..)

Tests in error: 
  test_006_ReplaceFileGood(edu.harvard.iq.dataverse.api.FilesIT): Cannot get property 'files' on null object
  testForceReplace(edu.harvard.iq.dataverse.api.FilesIT): Cannot get property 'files' on null object

Tests run: 72, Failures: 8, Errors: 2, Skipped: 0

Also seeing exceptions on the API import DS good DOI path; but suspecting that's something where an API has changed it's response and the prototype migration scripts haven't keep up.  Will troubleshoot some more.

pameyer commented:
[server-20180611.txt](https://github.com/IQSS/dataverse/files/2091668/server-20180611.txt)
Initial thinking on exception traces (problem with expected response / downstream script parsing) was incorrect.  The script was unable to parse the HTML "internal server error" page as JSON.

michbarsinai commented:
There are multiple issues here. One of them might be that you need to change the settings for your Dataverse installation to support the new PID system. Here are my relevant settings:

```
       name        |               content                
-------------------+--------------------------------------
 :Protocol         | doi
 :DoiProvider      | EZID
 :Authority        | 10.5072
 :Shoulder         | FK2/

```

> On 11 Jun 2018, at 23:43, pameyer <notifications@github.com> wrote:
> 
> server-20180611.txt
> Initial thinking on exception traces (problem with expected response / downstream script parsing) was incorrect. The script was unable to parse the HTML "internal server error" page as JSON.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

oscardssmith commented on src/main/java/edu/harvard/iq/dataverse/engine/command/impl/UpdateDatasetVersionCommand.java:
```diff
@@ -66,9 +66,16 @@ public Dataset execute(CommandContext ctxt) throws CommandException {
         validateOrDie( getDataset().getEditVersion(), isValidateLenient() );
         
-        final DatasetVersion editVersion = getDataset().getEditVersion();
+        DatasetVersion editVersion = getDataset().getEditVersion();
```
why not final? I don't see it being changed.

michbarsinai commented:
Update: All tests in the IT tests pass.

pdurbin commented:
@michbarsinai I took a look at this branch this morning.

First, the branch has merge conflicts for the following files:

- src/main/java/edu/harvard/iq/dataverse/Dataset.java
- src/main/java/edu/harvard/iq/dataverse/DatasetPage.java
- src/main/java/edu/harvard/iq/dataverse/api/Datasets.java
- src/main/java/edu/harvard/iq/dataverse/engine/command/impl/CreateDatasetCommand.java 

Can you please resolve these?

Second, not all of the API tests are passing on my laptop. One test is failing: Failed tests:   testSetCreation(edu.harvard.iq.da​taverse.api.HarvestingServerIT): expected:<201> but was:<500>

I dug into this a little and found this:

Caused by: java.lang.NullPointerException
	at edu.harvard.iq.dataverse.harvest.server.OAISetServiceBean.exportOaiSet(OAISetServiceBean.java:149)

On line 149 datasetIds is null:

![screen shot 2018-07-02 at 10 36 12 am](https://user-images.githubusercontent.com/21006/42170355-4f1ae3de-7de4-11e8-8391-c9e309e08466.png)

You're saying that this test is passing on your laptop?

I did make a couple commits to your branch to get the API test suite to execute on a fresh clone of the git repo at 6400958 and a84150d after consulting with @pameyer at http://irclog.iq.harvard.edu/dataverse/2018-07-02 . So please pull the latest. I don't think these will interfere with any of your work. Thanks!

michbarsinai commented:
Thanks. I've already resolved the merge conflicts (testing now). As for the tests - yes. Twice. I'll pull the latest and see how this goes.

> On 2 Jul 2018, at 17:41, Philip Durbin <notifications@github.com> wrote:
> 
> @michbarsinai <https://github.com/michbarsinai> I took a look at this branch this morning.
> 
> First, the branch has merge conflicts for the following files:
> 
> src/main/java/edu/harvard/iq/dataverse/Dataset.java
> src/main/java/edu/harvard/iq/dataverse/DatasetPage.java
> src/main/java/edu/harvard/iq/dataverse/api/Datasets.java
> src/main/java/edu/harvard/iq/dataverse/engine/command/impl/CreateDatasetCommand.java
> Can you please resolve these?
> 
> Second, not all of the API tests are passing on my laptop. One test is failing: Failed tests: testSetCreation(edu.harvard.iq.da​taverse.api.HarvestingServerIT): expected:<201> but was:<500>
> 
> I dug into this a little and found this:
> 
> Caused by: java.lang.NullPointerException
> at edu.harvard.iq.dataverse.harvest.server.OAISetServiceBean.exportOaiSet(OAISetServiceBean.java:149)
> 
> On line 149 datasetIds is null:
> 
>  <https://user-images.githubusercontent.com/21006/42170355-4f1ae3de-7de4-11e8-8391-c9e309e08466.png>
> You're saying that this test is passing on your laptop?
> 
> I did make a couple commits to your branch to get the API test suite to execute on a fresh clone of the git repo at 6400958 <https://github.com/IQSS/dataverse/commit/64009589304d155cdb06cb1cb431bb09d5524f3e> and a84150d <https://github.com/IQSS/dataverse/commit/a84150dc89d30099f2b36184324d9b2e35d9001b> after consulting with @pameyer <https://github.com/pameyer> at http://irclog.iq.harvard.edu/dataverse/2018-07-02 <http://irclog.iq.harvard.edu/dataverse/2018-07-02> . So please pull the latest. I don't think these will interfere with any of your work. Thanks!
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/IQSS/dataverse/pull/4606#issuecomment-401828758>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AB2UJEUWeipu-dtmkQFd5s8RTNycb_0Uks5uCjE3gaJpZM4Te5j5>.
>

pdurbin commented:
@michbarsinai thanks. Meanwhile I'm checking if the HarvestingServerIT failure is also on the "develop" branch.

michbarsinai commented:
Currently a single test is failing: `testAddUpdateDatasetViaNativeAPI`.

sekmiller commented:
I added that test for the edit dataset API which was recently merged with Dev. I'll take a look to see why it's failing.

pdurbin commented:
@sekmiller thanks

@michbarsinai are you using EZID or DataCite? I just opened #4800

michbarsinai commented:
Dunno. I’m using whatever is the integration test default (not near my laptop so can’t check). 

Sent from my iPhone

> On 2 Jul 2018, at 19:55, Philip Durbin <notifications@github.com> wrote:
> 
> @sekmiller thanks
> 
> @michbarsinai are you using EZID or DataCite? I just opened #4800
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

michbarsinai commented:
Thanks, @sekmiller.

Sent from my iPhone

> On 2 Jul 2018, at 18:43, Stephen Kraffmiller <notifications@github.com> wrote:
> 
> I added that test for the edit dataset API which was recently merged with Dev. I'll take a look to see why it's failing.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

pdurbin commented:
> testAddUpdateDatasetViaNativeAPI

@michbarsinai that test is failing on the "develop" branch too. I just opened #4803 about it.

michbarsinai commented:
Ok. I’ve fixed it in the 3083 branch: it was waiting for a wrong message (different case and missing colon).  I’ve added an equalsToCI method, for case insensitive comparison. 

Sent from my iPhone

> On 3 Jul 2018, at 18:58, Philip Durbin <notifications@github.com> wrote:
> 
> testAddUpdateDatasetViaNativeAPI
> 
> @michbarsinai that test is failing on the "develop" branch too. I just opened #4803 about it.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

oscardssmith commented on src/main/java/edu/harvard/iq/dataverse/engine/command/impl/AbstractCreateDatasetCommand.java:
```diff
@@ -0,0 +1,168 @@
+        registrationRequired = isRegistrationRequired;
+    }
+   
+    protected void additionalParameterTests(CommandContext ctxt) throws CommandException {
```
why not make this abstract?

michbarsinai replied:
Doing nothing is the default, valid behavior. So either have a an empty method in some subclasses, or have a single one here.

oscardssmith commented on src/main/java/edu/harvard/iq/dataverse/engine/command/impl/AbstractCreateDatasetCommand.java:
```diff
@@ -0,0 +1,168 @@
+     * @param ctxt 
+     * @throws edu.harvard.iq.dataverse.engine.command.exception.CommandException 
+     */
+    protected void postPersist( Dataset theDataset, CommandContext ctxt ) throws CommandException {
```
here too.

michbarsinai replied:
(same as above)

oscardssmith commented on src/main/java/edu/harvard/iq/dataverse/engine/command/impl/AbstractCreateDatasetCommand.java:
```diff
@@ -0,0 +1,168 @@
+        }
+        
+        String nonNullDefaultIfKeyNotFound = "";
+        if (theDataset.getProtocol()==null) theDataset.setProtocol(ctxt.settings().getValueForKey(SettingsServiceBean.Key.Protocol, nonNullDefaultIfKeyNotFound));
```
indentation and brakcets please

michbarsinai replied:
Done. (we might need to edit the coding guidelines, but agreed that for now we need indentation and brackets).

matthew-a-dunlap commented on src/main/java/edu/harvard/iq/dataverse/GlobalId.java:
```diff
@@ -119,9 +149,11 @@ public URL toURL() {
+     *         {@code null} if parsing failed.
      */
-    
     private boolean parsePersistentId(String identifierString) {
```
As part of #4761 we have been discussing pid structure and the challenges of external systems using these values. There is concern that we will start importing pids with a lot of weird characters, for example this valid DOI:

`https://doi.org/10.1002/(sici)1099-1409(199908/10)3:6/7<672::aid-jpp192>3.0.co;2-8`

This is of more concern as we are trying to have all external systems use pids to reference files and datasets, but these edge case pids could lead to systems breaking, injection attacks, etc.

A solution that was discussed was to enforce limits on the characters we import. The general idea is to be pretty strict but simple and then add in edge cases as requested. 

See here for more info on DataCite's pid. I did not read into Handle's or ezid, tho my limited understanding is that those allow a different/weird assortment of characters. https://support.datacite.org/docs/doi-basics https://www.crossref.org/blog/dois-and-matching-regular-expressions/

matthew-a-dunlap replied:
@michbarsinai I'm worried about "opening the floodgates" and allowing in pids well outside our normal structure. My understanding is that before this PR it was impossible to import DOIs in. It would be much better to have the restrictions from the get-go. 

Maybe this could be done in another issue before the next release as the work is small and fairly isolated. @scolapasta thoughts?

scolapasta replied:
Yes, let's keep it separate. @matthew-a-dunlap can you open an issue? We can convince @djbrooke of its priority next week when he's back.

matthew-a-dunlap replied:
@scolapasta will do before I head out!

michbarsinai replied:
I defer to @scolapasta. Here are my thought on this: We are parsing the PID using `GlobalId.parse` (https://github.com/IQSS/dataverse/blob/3083-support-migration-with-dois/src/main/java/edu/harvard/iq/dataverse/api/Dataverses.java#L257)
So changes there will be used in the new code here. Also, this is a large PR (131 files) so any delay here causes a lot of merge and delete/edit conflicts. So I think implementing #4761 in a new PR would make more sense.

michbarsinai commented on src/main/java/edu/harvard/iq/dataverse/engine/command/impl/UpdateDatasetVersionCommand.java:
```diff
@@ -1,85 +1,151 @@
  */
 @RequiredPermissions(Permission.EditDataset)
-public class UpdateDatasetVersionCommand extends AbstractCommand<DatasetVersion> {
+public class UpdateDatasetVersionCommand extends AbstractDatasetCommand<Dataset> {
```
No. It modified the only edit version of a dataset, so in fact works on the dataset. It makes even more sense when you look at the sequence of commands (createXXXDataset -> EditDatasetVersion->PublishDataset).

We can discuss the whole separation between Dataset and DatasetVersion, but it's beyond the scope of this PR.

michbarsinai commented on src/main/java/edu/harvard/iq/dataverse/engine/command/impl/CreateNewDatasetCommand.java:
```diff
@@ -0,0 +1,100 @@
+         // dataset returned by the Create command. That issue was fixed by adding 
+         // a full refresh of the datast with datasetService.find() between the 
+         // two commands. But it may be a good idea to make sure they are properly
+         // linked here (?)
```
Future, if any.

michbarsinai commented on src/main/java/edu/harvard/iq/dataverse/engine/command/impl/AbstractDatasetCommand.java:
```diff
@@ -0,0 +1,207 @@
+public abstract class AbstractDatasetCommand<T> extends AbstractCommand<T> {
+
+    private static final Logger logger = Logger.getLogger(AbstractDatasetCommand.class.getName());
+    private static final int FOOLPROOF_RETRIAL_ATTEMPTS_LIMIT = 2 ^ 8;
```
And an old one too (part of an old code that was migrated here). But it makes sense and is pretty clear and also rather funny, so I think we should keep it :-)

michbarsinai commented on src/main/java/edu/harvard/iq/dataverse/engine/command/impl/AbstractDatasetCommand.java:
```diff
@@ -0,0 +1,207 @@
+
+    private static final Logger logger = Logger.getLogger(AbstractDatasetCommand.class.getName());
+    private static final int FOOLPROOF_RETRIAL_ATTEMPTS_LIMIT = 2 ^ 8;
+    private Dataset dataset;
```
Design choice. `protected` allows future implementation direct access to this field, which allows them to break things in subtle ways (as we give away the integrity of this class' internal state to the implementation). It's a Dataset command, so there should be a dataset involved, hence it makes sense to have one.

oscardssmith replied:
My thinking was that since it's a `Dataset` command it makes sense that implementations have access to the dataset without going through a wrapper.

pdurbin commented:
> that test is failing on the "develop" branch too. I just opened #4803 about it.

I merged #4804 and now the API test suite is passing on "develop" for the first time in time in a week and a half.
